x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java

```diff
@@ -246,25 +266,43 @@ public void refreshJobMemory(String jobId, ActionListener<Long> listener) {
             return;
         }
 
+        // The phaser prevents searches being started after the memory tracker's stop() method has returned
+        if (stopPhaser.register() != 0) {
```

Comment:
Is it possible for `refreshJobMemory` to be called concurrently? 

If so, doesn't that mean that we could have one register, increment the registration phase, and then another arrive. That second one that arrived will not execute as its arrival is > 0, even though we are not stopping. 

I may be misunderstanding how the phaser works.

Reply:
> Is it possible for `refreshJobMemory` to be called concurrently?

Yes.

> That second one that arrived will not execute as its arrival is > 0, even though we are not stopping.

[`Phaser.register()`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Phaser.html#register--) increments the _number of parties_ but returns the _phase number_.  For the ML memory tracker the phase number is the number of times `stop()` has been called.

The number of parties to the phaser is 1 plus the number of concurrent calls to `refreshJobMemory()`.  There will only be a problem if there are 2^31 concurrent calls, as this will overflow the number of parties.  I'm assuming that will never happen.

Reply:
I was also concerned about this. I wonder if it would be simpler to have a volatile boolean to mark that `stop` was called,  ensure no new search requests are accepted when that boolean is true and let searches always register/deregister to the phaser.

Reply:
Ah, ok. Your explanation covers me. I also assumed `register()` returns the number of parties, not the phase. Makes sense.

Reply:
I grok now. 👍